MetaMask/metamask-mobile
# feat: Add RPC Method Tracking to Dapp JSON RPC Engine pipeline

## **Description**

Currently in Mobile, we do not directly track RPC Method calls in metrics. It would be great to have the same insights on RPC method calls that we currently do in Extension. This PR adds a simplified version of the [RPCMethodTrackingMiddleware found in Extension](https://github.com/MetaMask/metamask-extension/blob/0cd76e5b6a3fe4b491348ac0c348a27a82368a32/app/scripts/lib/createRPCMethodTrackingMiddleware.js#L129) to the Mobile Dapp JSON RPC Engine pipeline.

## **Related issues**

Fixes: https://github.com/MetaMask/MetaMask-planning/issues/2169

## **Manual testing steps**

1. Use Flipper or some other software that monitors app console logs and network events
2. Start the iOS simulator build
3. Open a Dapp (such as the [test dapp](https://metamask.github.io/test-dapp/)) in the  in-app browser
4. Use the Dapp
5. Filter the app console logs for `TRACK event saved` AND/OR Check the network tab for `Inpage Provider`: `Provider Method Called` events that include the method name and origin
  - methods like `eth_sign` should not be rate limited
  - methods like `eth_requestAccounts` should be rate limited one tracked event every 5 minutes
  - most other methods take on the default behavior of being sampled randomly with a 0.1% chance, meaning it will be very difficult to get these ones to trigger events 
  - finally, there is a max 10 events per sliding 5 minute window that all events above are gated by


## **Screenshots/Recordings**



### **Before**



### **After**


https://github.com/MetaMask/metamask-mobile/assets/918701/4bef5012-894b-4112-b6a3-6b83f91a3a1b



## **Pre-merge author checklist**


## **Pre-merge reviewer checklist**
